sirepo/request.py

```diff
@@ -18,38 +19,116 @@
 
 _SIM_TYPE_ATTR = "sirepo_http_request_sim_type"
 
+#: We always use the same name for a file upload
+_FORM_FILE_NAME = "file"
 
-def init_quest(qcall):
+
+def init_quest(qcall, internal_req=None):
     if qcall.bucket_unchecked_get("in_srunit"):
         sreq = _SRequest(
             http_authorization=None,
             http_headers=PKDict(),
             http_method="GET",
             http_request_uri="/",
-            http_server_uri="http://localhost",
-            internal_req=None,
+            http_server_uri="http://localhost/",
+            internal_req=internal_req,
             remote_addr="0.0.0.0",
         )
-    else:
+    elif "werkzeug" in str(type(internal_req)):
         import flask
 
         sreq = _SRequest(
-            http_authorization=flask.request.authorization,
-            http_headers=flask.request.headers,
-            http_method=flask.request.method,
-            http_request_uri=flask.request.url,
+            http_authorization=internal_req.authorization,
+            http_headers=internal_req.headers,
+            http_method=internal_req.method,
+            http_request_uri=internal_req.url,
             http_server_uri=flask.url_for("_flask_dispatch_empty", _external=True),
-            internal_req=flask.request,
-            remote_addr=flask.request.remote_addr,
+            internal_req=internal_req,
+            remote_addr=internal_req.remote_addr,
+            _body_as_bytes=lambda: internal_req.get_data(cache=False),
+            _form_file_class=_FormFileFlask,
+            _form_get=internal_req.form.get,
         )
+    elif "tornado" in str(type(internal_req)):
+        r = internal_req.request
+        sreq = _SRequest(
+            http_authorization=_parse_authorization(r.headers.get("Authorization")),
+            http_headers=r.headers,
+            http_method=r.method,
+            http_request_uri=r.path,
+            http_server_uri=f"{r.protocol}://{r.host}/",
+            internal_req=internal_req,
+            remote_addr=r.remote_ip,
+            _body_as_bytes=lambda: internal_req.request.body,
```

Comment:
Why are these public and the other fields aren't? Why does something like http_authorization not need an accessor method but  body_as_bytes does?

Reply:
> Why are these public and the other fields aren't? 

_body_as_bytes, etc. are only referenced inside request.py. The others are accessed outside request.py.

> Why does something like http_authorization not need an accessor method but body_as_bytes does?

body_as_bytes could be an accessor. A refactoring that could be done, but I'll leave as now. Perhaps documents better.

I like that _parse_authorization is always called to validate the header on every request at request time (before other code gets to execute). flask.request.authorization validates at request startup, too. 

I was striking a balance between too much wrapping so I could get things done. In general internal_req should be used sparingly and that's why it is called awkwardly internal_req.

Reply:
I simplied body_as_bytes d57937d0

Reply:
> _body_as_bytes, etc. are only referenced inside request.py. The others are accessed outside request.py.

I believe `http_headers` is only accessed inside request.py. The others are accessed outside but through a method that does nothing else than call `self._<name>()`. So, I'm still not sure what the difference is between the two? Why not have them all be public or private?

> In general internal_req should be used sparingly

It would be helpful to document this somewhere in the code.

Reply:
Sorry I left my review before seeing your update. But I think my points still stand.

Reply:
http_headers should be made private. `http_headers_uget` is not the same `body_as_bytes`. In this case, there'd be lambda's in three places that would do the gets. `_http_headers` implements `get`, but the object is different in the three cases.